certdb/certdb.go

```diff
@@ -29,252 +25,16 @@ type OCSPRecord struct {
-	if err != nil {
-		return nil, wrapCertStoreError(err)
-	}
-	defer rows.Close()
-
-	if rows.Next() {
-		return rr, sqlstruct.Scan(rr, rows)
-	}
-	return nil, nil
-}
-
-// GetUnexpiredOCSPs retrieves all unexpired OCSPRecord from db.
-func GetUnexpiredOCSPs(db *sql.DB) (rrs []*OCSPRecord, err error) {
-	rr := new(OCSPRecord)
-	rows, err := db.Query(fmt.Sprintf(selectAllUnexpiredOCSPSQL, sqlstruct.Columns(*rr)))
-	if err != nil {
-		return nil, wrapCertStoreError(err)
-	}
-	defer rows.Close()
-
-	for rows.Next() {
-		err = sqlstruct.Scan(rr, rows)
-		if err != nil {
-			return nil, wrapCertStoreError(err)
-		}
-		rrs = append(rrs, rr)
-	}
-
-	return rrs, nil
-}
-
-// UpdateOCSP updates a ocsp response record with a given serial number.
-func UpdateOCSP(db *sql.DB, serial, body string, expiry time.Time) (err error) {
-	var result sql.Result
-	result, err = db.Exec(updateOCSPSQL, serial, body, expiry)
-
-	if err != nil {
-		return wrapCertStoreError(err)
-	}
-
-	var numRowsAffected int64
-	numRowsAffected, err = result.RowsAffected()
-
-	if numRowsAffected == 0 {
-		return cferr.Wrap(cferr.CertStoreError, cferr.RecordNotFound, fmt.Errorf("failed to update the OCSP record"))
-	}
-
-	if numRowsAffected != 1 {
-		return wrapCertStoreError(fmt.Errorf("%d rows are affected, should be 1 row", numRowsAffected))
-	}
-	return
-}
-
-// UpsertOCSP update a ocsp response record with a given serial number,
-// or insert the record if it doesn't yet exist in the db
-// Implementation note:
-// We didn't implement 'upsert' with SQL statement and we lost race condition
-// prevention provided by underlying DMBS.
-// Reasoning:
-// 1. it's diffcult to support multiple DBMS backends in the same time, the
-// SQL syntax differs from one to another.
-// 2. we don't need a strict simultaneous consistency between OCSP and certificate
-// status. It's OK that a OCSP response still shows 'good' while the
-// corresponding certificate is being revoked seconds ago, as long as the OCSP
-// response catches up to be eventually consistent (within hours to days).
-// Write race condition between OCSP writers on OCSP table is not a problem,
-// since we don't have write race condition on Certificate table and OCSP
-// writers should periodically use Certificate table to update OCSP table
-// to catch up.
-func UpsertOCSP(db *sql.DB, serial, body string, expiry time.Time) (err error) {
-	var result sql.Result
-	result, err = db.Exec(updateOCSPSQL, serial, body, expiry)
-
-	if err != nil {
-		return wrapCertStoreError(err)
-	}
-
-	var numRowsAffected int64
-	numRowsAffected, err = result.RowsAffected()
-
-	if numRowsAffected == 0 {
-		return InsertOCSP(db, &OCSPRecord{Serial: serial, Body: body, Expiry: expiry})
-	}
-
-	if numRowsAffected != 1 {
-		return wrapCertStoreError(fmt.Errorf("%d rows are affected, should be 1 row", numRowsAffected))
-	}
-	return
+// DBAccessor abstracts the CRUD of certdb objects from a DB.
+type DBAccessor interface {
+	SetDB(db *sql.DB)
```

Comment:
I think it would probably be cleaner to not have `SetDB` be a property of the `DBAccessor` interface, but instead have types that satisfy `DBAccessor` take care of setting the DB as part of their constructor.